### PR TITLE
Test refactor: Query operations

### PR DESCRIPTION
**Pull request by octaviansima**
Part 1 of 2 of the new testing infrastructure. Part 2 will include all Opaque functionality that cannot be tested against Spark (EncryptedSource, etc.)

This PR includes new tests for all the existing query operations that are supported, as well as those that are not yet supported (ignored). It also refactors much of the existing infrastructure, and defines new base classes that simplify creating tests for the plain DataFrame API as well as SQL queries. All queries are taken directly from Spark's own testing.

### CODE REVIEW

File: src/test/scala/edu/berkeley/cs/rise/opaque/AggregationSuite.scala [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT avg(value) FROM agg1
+        """.stripMargin)
+    }
+  }
+
+  ignore("first_value and last_value") {
```
Comment (wzheng) [inline src/test/scala/edu/berkeley/cs/rise/opaque/AggregationSuite.scala:233]: Will we have support for this in the future?

File: src/test/scala/edu/berkeley/cs/rise/opaque/JoinSuite.scala [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+  test("inner join, no matches") {
+    checkAnswer() { sl =>
+      val x = testData2(sl).where($"a" === 1).as("x")
+      val y = testData2(sl).where($"a" === 2).as("y")
+      x.join(y).where($"x.a" === $"y.a")
+    }
+  }
+
+  ignore("big inner join, 4 matches per row") {
```
Comment (wzheng) [inline src/test/scala/edu/berkeley/cs/rise/opaque/JoinSuite.scala:88]: Why is this ignored?

File: src/test/scala/edu/berkeley/cs/rise/opaque/QEDSuite.scala [hunk 3]
```diff
@@ -18,22 +18,14 @@
 package edu.berkeley.cs.rise.opaque
```
Comment (wzheng) [inline src/test/scala/edu/berkeley/cs/rise/opaque/QEDSuite.scala:18]: Maybe we should rename `QEDSuite` -- it was an old name for Opaque :P

File: src/test/scala/edu/berkeley/cs/rise/opaque/AggregationSuite.scala (hunk 1 above)
Comment (octaviansima) [reply src/test/scala/edu/berkeley/cs/rise/opaque/AggregationSuite.scala:233]: I would think so? Seems like a function that is supported by Spark SQL https://spark.apache.org/docs/3.1.1/api/sql/index.html#first_value

File: src/test/scala/edu/berkeley/cs/rise/opaque/JoinSuite.scala (hunk 2 above)
Comment (octaviansima) [reply src/test/scala/edu/berkeley/cs/rise/opaque/JoinSuite.scala:88]: Casting error :( 
`java.lang.ClassCastException: org.apache.spark.sql.execution.UnionExec cannot be cast to edu.berkeley.cs.rise.opaque.execution.OpaqueOperatorExec`
Should be a supported query, but looks we need to fix some bugs.

**Pull request merged**